hudi-client/hudi-spark-client/src/test/java/org/apache/hudi/table/functional/TestCleanPlanExecutor.java

```diff
@@ -683,4 +694,56 @@ public void testKeepXHoursWithCleaning(
     assertFalse(testTable.baseFileExists(p0, firstCommitTs, file1P0C0));
     assertFalse(testTable.baseFileExists(p1, firstCommitTs, file1P1C0));
   }
+
+  @Test
+  public void testGetEarliestCommitToRetain() {
+    HoodieWriteConfig config = HoodieWriteConfig.newBuilder()
+            .withPath(basePath)
+            .withSchema(HoodieTestDataGenerator.TRIP_EXAMPLE_SCHEMA)
+            .withMetadataConfig(HoodieMetadataConfig.newBuilder()
+                    .withAssumeDatePartitioning(true)
+                    .build())
+            .withAutoCommit(false)
+            .withCleanConfig(HoodieCleanConfig.newBuilder()
+                    .withIncrementalCleaningMode(true)
+                    .withFailedWritesCleaningPolicy(HoodieFailedWritesCleaningPolicy.LAZY)
+                    .withCleanerPolicy(HoodieCleaningPolicy.KEEP_LATEST_COMMITS)
+                    .retainCommits(5)
+                    .build())
+            .build();
+    SparkRDDWriteClient writeClient = getHoodieWriteClient(config);
+    IntStream.rangeClosed(1, 9).mapToObj(i -> {
+      String newCommitTime = "00" + i;
+      List<HoodieRecord> records = dataGen.generateInserts(newCommitTime, 10);
+      JavaRDD<HoodieRecord> writeRecords = jsc.parallelize(records, 1);
+      writeClient.startCommitWithTime(newCommitTime);
+      JavaRDD<WriteStatus> writeStatues = writeClient.insert(writeRecords, newCommitTime);
+      // Assuming the first commit is pending, simulating the situation where all instants before the first pending commit have been achieved.
```

Comment:
Your understanding of the premise is correct, but a pending commit may be caused by delayed heartbeats or the execution of a replace commit that takes a long time, resulting in many commits on the timeline. In this case, I think it is reasonable for the archive to remove all commits before this pending commit. However, this can lead to the problem I mentioned, that is, when performing incremental clean, the endpoint of the last clean record cannot be found as the starting point of this clean, which will result in a full clean. If we mark the pending commit record as the endpoint of this case in getPartitionPathsForIncrementalCleaning, we can find a timeline greater than or equal to the starting point and less than the endpoint, which means that the pending commit will not be removed. This operation maintains the original intention without changing other situations and solves this problem. I'm not sure if I've expressed my understanding clearly, do you think what I said makes sense?

Reply:
So, in the next incremental clean, if this clean is still pending, neither the starting nor the ending point of this clean will have any commits cleaned. However, if this pending commit is rolled back or executed, then the starting point will be this pending commit and the ending point will be the normal commit. In this case, the clean will be executed normally and no files will be missed.

Reply:
> In this case, I think it is reasonable for the archive to remove all commits before this pending commit

I don't think this is reasonable. 

```
  /**
   * A FileSlice is considered committed, if one of the following is true - There is a committed data file - There are
   * some log files, that are based off a commit or delta commit.
   */
  private boolean isFileSliceCommitted(FileSlice slice) {
    if (!compareTimestamps(slice.getBaseInstantTime(), LESSER_THAN_OR_EQUALS, lastInstant.get().getTimestamp())) {
      return false;
    }

    return timeline.containsOrBeforeTimelineStarts(slice.getBaseInstantTime());
  }
```

Given your above assumption, a `fileSlice` that may be the output of a inflight replacecommit/clustering will be flagged as committed, when in actuality, it is not...